byterun/intern.c

```diff
@@ -718,15 +720,25 @@ value caml_input_val(struct channel *chan)
   }
   /* Initialize global state */
   intern_init(block, block);
-  intern_alloc(h.whsize, h.num_objects);
+  intern_alloc(h.whsize, h.num_objects, outside_heap);
   /* Fill it in */
   intern_rec(&res);
-  intern_add_to_heap(h.whsize);
+  if (!outside_heap) {
+    intern_add_to_heap(h.whsize);
+  } else {
+    intern_extra_block = NULL;
+    intern_block = 0;
+  }
```

Comment:
I don't understand your comment (or else I am misunderstanding something).  We clearly cannot free the block there since it contains the demarshalled data.  There is no support at present for recovering this memory at all; we may implement support for that subsequently.  I'm not worried about the calling of caml_alloc_for_heap without calling the add or free functions.  We could change the comment in memory.c to reflect that this is safe.